### ISSUE TITLE
Backward with Riemannian Gradient

### ISSUE THREAD

**Issue by ShaoTengLiu** [geoopt/geoopt#98]
Hi, I wondered how does your loss.backward() work with Riemannian Gradient. I think the egrad2rgrad() may be an essential function, but it seems it is not used in any code. Do we need to extend torch.autograd by Function? I mean redefine the grad() and backward()?

The reason I ask this question is that I tried to use MobiusLinear, but it is easy for the tensor to be nan. I believe the problem is that the weight of the model becomes nan when loss.backward() is called.

Thanks a lot!

**Comment by ferrine** [comment]
Nan's are related to numerical stability in forward / backward pass. You a should take care to keep points inside the boundary. The easiest way is to use math methods of the manifold instance, they do projection automatically.

**Comment by newkozlukov** [comment]
@ShaoTengLiu after the `.backward()` the `.grad`s of parameters contain what authors of reference paper refer to as "Euclidean gradients"; they, apparently, _encode_ (see 1)  a _co_-tangent vector on the manifold (see 2), and prior to usage they need be converted to tangent _vectors_ (see 3) (for our needs, it's something representing a "velocity" that you can feed into exponential map and parallel transport) via `egrad2rgrad` function; you can see an example of using these gradients in https://github.com/geoopt/geoopt/blob/a544c6ffe39478b3829a3bd7d065cdc45dc2e98b/geoopt/optim/rsgd.py#L93

This is not a very intuitive design, but that's about what's available without penetrating pytorch's source tree (which is the only way to extend `Tensor` at the time). We're reluctant to wrap tensors into pure-python proxy objects because of performance concerns.

The implementation's due to @ferrine, I hope he may correct me if I missed anything.

* * *

1. By "encodes" I mean that for each implemented manifold there is an implicit immersion of that manifold into some Euclidean (usually ambient) space, e.g. $\mathbb{R}^3$ for a 2-sphere. So, as far as I understand, the `backward()`'s `.grad`s must be cotangent vectors corresponding to pushforwards of actual gradients with respect to that immersion. So, our understanding is that for each manifold there's an immersion, and so the entire product manifold of all parameters may also be embedded into product of Euclidean spaces and that's where `pytorch` operates, with all manifold-to-manifold mappings being implicitly encoded by their compositions with immersions and inverse immersions (so that we end up with mappings between Euclidean spaces, taking and producing numerical arrays).
2. Reverse-mode autodiff basically performs a pullback of a cotangent vector (see Betancourt). So `grad`s must contain coordinates of a cotangent vector in the ambient space (cotangent is a linear functional on tangent space -- takes a tangent vector, produces a number). To obtain the actual tangent vector that we feed into `expmap` we would pull that cotangent back (obtaining a cotangent on the manifold) using immersion, then apply inverse metric tensor (which would give us the tangent vector inner product with which is equivalent to applying the cotangent). However, we never access the actual tangents or cotangents or even points of manifold in code, we only work with their numarray-representations, produced by immersion. So the result of `egrad2rgrad` is pushforward of that actual direction.
3. Again, tangent vectors are represented by their pushforwards wrt immersion

We're still in the process of comprehending and formalizing this all though -- so we would be glad to hear any related thoughts. I'm not even sure if I used the word "immersion" correctly, it may be some other structure.

**Comment by ShaoTengLiu** [comment]
Thanks for your careful reply. It helps me a lot.
In addition, I finally find that my problem is caused by this code: 
x = x.clamp(-1 + 1e-15, 1 - 1e-15) in Artanh defined in math.py
This threshold is not always safe. For my data, it is more proper to set it to 1e-5
Thanks again!
